components/homme/src/share/cxx/utilities/Hash.hpp

```diff
@@ -0,0 +1,69 @@
+/********************************************************************************
+ * HOMMEXX 1.0: Copyright of Sandia Corporation
+ * This software is released under the BSD license
+ * See the file 'COPYRIGHT' in the HOMMEXX/src/share/cxx directory
+ *******************************************************************************/
+
+#ifndef HOMMEXX_HASH_HPP
+#define HOMMEXX_HASH_HPP
+
+#include <cstdint>
+
+#include "Types.hpp"
+
+/* Utilities to calculate a hash for a given model state. Hash values can be
+   compared between runs to find instances of non-BFBness.
+ */
+
+namespace Homme {
+
+typedef std::uint64_t HashType;
+
+// Each hash function accumulates v into accum using a hash of its bits.
+
+KOKKOS_INLINE_FUNCTION void hash (const HashType v, HashType& accum) {
+  constexpr auto first_bit = 1ULL << 63;
+  accum += ~first_bit & v; // no overflow
+  accum ^=  first_bit & v; // handle most significant bit  
+}
```

Comment:
Is the hash a simple bitwise accumulation or did I miss something?
Trying to understand potential for hash collisions.

Reply:
It's just double -> unsigned long long, then add these. I handle the most significant bit manually so there isn't an overflow if overflow detection is on. -x + x isn't 0 (as with signed add) and x + x isn't 0 (as with XOR), so it's good enough for non-BFB detection.